pkg/file/file.go

```diff
@@ -193,3 +196,86 @@ func ExtractFilenameFromURLPath(p string, defaultExtension string) string {
 
 	return b
 }
+
+// ExtractTarGz extracts the contents of the provided tar.gz file.
+// This implementation uses os.* calls without permission checks, as we're
+// running this operation in the context of fleetd running as root on a host
+// (e.g. for installs), so we have different constraints than fleetctl building
+// a package. destDir should be provided by the code rather than user input to
+// avoid directory traversal attacks. maxFileSize indicates how large we want
+// to allow the max file size to be when decompressing, as a zip bomb mitigation.
+func ExtractTarGz(path string, destDir string, maxFileSize int64) error {
+	tarGzFile, err := os.Open(path)
+	if err != nil {
+		return fmt.Errorf("open %q: %w", path, err)
+	}
+	defer tarGzFile.Close()
+
+	gzipReader, err := gzip.NewReader(tarGzFile)
+	if err != nil {
+		return fmt.Errorf("gzip reader: %w", err)
+	}
+	defer gzipReader.Close()
+
+	tarReader := tar.NewReader(gzipReader)
+
+	for {
+		header, err := tarReader.Next()
+		switch {
+		case err == nil:
+			// OK
+		case errors.Is(err, io.EOF):
+			return nil
+		default:
+			return fmt.Errorf("tar reader: %w", err)
+		}
+
+		// Prevent zip-slip attack (which, combined with a trusted destDir, remediates the potential directory traversal
+		// attack below)
+		if strings.Contains(header.Name, "..") {
+			return fmt.Errorf("invalid path in tar.gz: %q", header.Name)
+		}
+
+		targetPath := filepath.Join(destDir, header.Name) // nolint:gosec // see above notes on dir traversal
+
+		switch header.Typeflag {
+		case tar.TypeDir:
+			if err := os.MkdirAll(targetPath, constant.DefaultDirMode); err != nil {
+				return fmt.Errorf("mkdir %q: %w", header.Name, err)
+			}
+		case tar.TypeReg:
+			err := func() error {
+				outFile, err := os.OpenFile(targetPath, os.O_CREATE|os.O_WRONLY, header.FileInfo().Mode())
+				if err != nil {
+					return fmt.Errorf("failed to create %q: %w", header.Name, err)
+				}
+				defer outFile.Close()
+
+				// CopyN call to avoid zip bomb DoS since we have less control over arbitrary .tar.gz archives
+				// than in e.g. a TUF case.
+				var readBytes int64
+				chunkSize := int64(65536) // 64KiB
+				for {
+					if readBytes+chunkSize > maxFileSize {
+						return fmt.Errorf("aborted extraction of oversized file after %d bytes", readBytes)
+					}
+
+					_, err := io.CopyN(outFile, tarReader, chunkSize)
+					if err != nil {
+						if err == io.EOF {
+							break
+						}
+						return fmt.Errorf("failed to extract file %q inside %q: %w", header.Name, path, err)
+					}
```

Comment:
Missing updating `readBytes` after a successful read.

Reply:
Thanks and sorry for the miss. Fixing shortly.

Reply:
Fixed.